axiom/nr_php_packages.c

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright 2022 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <math.h>
+#include <stdlib.h>
+#include <stdio.h>
+
+#include "nr_axiom.h"
+#include "nr_limits.h"
+#include "nr_php_packages.h"
+#include "util_memory.h"
+#include "util_minmax_heap.h"
+#include "util_vector.h"
+#include "util_hashmap.h"
+#include "util_hashmap_private.h"
+#include "util_strings.h"
+
+nr_php_package_t* nr_php_package_create(char* name, char* version) {
+  nr_php_package_t* p = NULL;
+
+  if (NULL == name) {
+    return NULL;
+  }
+
+  p = (nr_php_package_t*)nr_malloc(sizeof(nr_php_package_t));
+  if (NULL == p) {
+    return NULL;
+  }
+
+  p->package_name = nr_strdup(name);
+  if (NULL != version) {
+    p->package_version = nr_strdup(version);
+  } else {
+    p->package_version
+        = nr_strdup(" ");  // if null, version is set to an empty
+                           // string with a space according to spec
+  }
+
+  return p;
+}
+
+void nr_php_package_destroy(nr_php_package_t* p) {
+  if (NULL != p) {
+    nr_free(p->package_name);
+    nr_free(p->package_version);
+    nr_free(p);
+  }
+}
+
+void nr_php_packages_add_package(nr_php_packages_t** h, nr_php_package_t* p) {
+  nr_php_package_t* package;
+  if (NULL == h) {
+    return;
+  }
+
+  if (NULL == p || NULL == p->package_name || NULL == p->package_version) {
+    return;
+  }
+
+  if (NULL == *h) {
+    *h = nr_hashmap_create((nr_hashmap_dtor_func_t)nr_php_package_destroy);
+  }
+
+  // If package with the same key already exists, we will check if the value is
+  // different. If it is different, then we will update the value of the package
+  package = (nr_php_package_t*)nr_hashmap_get(*h, p->package_name,
+                                              nr_strlen(p->package_name));
+  if (NULL != package) {
+    if (0 != nr_strcmp(package->package_version, p->package_version)) {
+      nr_free(package->package_version);
+      package->package_version = nr_strdup(p->package_version);
+    }
+    nr_php_package_destroy(p);
+    return;
+  }
+
+  nr_hashmap_set(*h, p->package_name, nr_strlen(p->package_name), p);
+}
+
+size_t nr_hashmap_count_buckets(const nr_php_packages_t* hashmap) {
```

Comment:
Why is this function here?  It is an internal function from `util_hashmap.c` and should not be duplicated this way!

Reply:
I have removed this function: [b4e182e](https://github.com/newrelic/newrelic-php-agent/pull/754/commits/b4e182ea97bab700def5aa6c9e1943d8cc04fdf1)